mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+
+        self._metric_info_artifact_need_update[run_id] = True
+        return metric_key
+
+    def register_metric_api_call(self, metric_name, call_command, call_pos_args, call_kwargs):
+        """
+        Given a metric api call (include the called metric function, and call arguments)
+        Register the call information (arguments dict) into the `metric_api_call_arg_dict_list_map`
+        and return a tuple of (is_ok, run_id, metric_key)
+        - metric key format is: {eval_dataset_name}_{metric_name}
+          metric_name is generated by metric function name, if there're multiple calls with
+          different call arguments, then appending index to the name.
+        """
+        call_arg_list = list(call_pos_args) + list(call_kwargs.values())
+
+        dataset_id_list = self._pred_result_id_to_dataset_name_and_run_id.keys()
+
+        # Note: some metric API the arguments is not like `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        for arg in call_arg_list:
+            if arg is not None and not np.isscalar(arg) and id(arg) in dataset_id_list:
+                dataset_name, run_id = self._pred_result_id_to_dataset_name_and_run_id[id(arg)]
+                break
+        else:
+            return False, None, None
+
+        metric_key = self.register_metric_info(
+            run_id, metric_name, dataset_name, call_command
+        )
+
+        return True, run_id, metric_key
+
+    def log_eval_metric(self, run_id, key, value):
+        # Note: if the case log the same metric key multiple times,
+        #  newer value will overwrite old value
+        client = mlflow.tracking.MlflowClient()
+        client.log_metric(
+            run_id=run_id,
+            key=key,
+            value=value
+        )
+        if self._metric_info_artifact_need_update[run_id]:
+            call_commands_list = []
+            for v in self._metric_api_call_info[run_id].values():
+                call_commands_list.extend(v)
+
+            call_commands_list.sort(key=lambda x: x[0])
+            dict_to_log = OrderedDict(call_commands_list)
+            client.log_dict(
+                run_id=run_id, dictionary=dict_to_log, artifact_file='metric_info.json'
+            )
+            self._metric_info_artifact_need_update[run_id] = False
+
+    def get_method_first_arg_value(self, method, call_pos_args, call_kwargs):
+        """
+        Get instance method first argument value (exclude the `self` argument).
+        """
+        if len(call_pos_args) >= 1:
+            return call_pos_args[0]
+        else:
+            first_arg_name = list(inspect.signature(method).parameters.keys())[1]
+            return call_kwargs.get(first_arg_name)
```

Comment:
Nit: This seems like a more general utility that isn't necessarily specific to metrics. Can we move it out of `_AutologTrainingStatus` and put it here: https://github.com/mlflow/mlflow/blob/9bd64dc0ec6a905c5d7a3d6b9855626ff9d2e0aa/mlflow/utils/autologging_utils/__init__.py#L96?